tests/unit/s2n_key_update_threads_test.c

```diff
@@ -0,0 +1,247 @@
+     * This proves that at least one KeyUpdate was sent.
+     */
+    RESULT_ENSURE_GT(conn->wire_bytes_out, S2N_DEFAULT_RECORD_LENGTH * seq_num);
+
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT s2n_test_encryption_limits(struct s2n_connection *conn)
+{
+    RESULT_GUARD_POSIX(s2n_example_negotiate(conn));
+
+    struct s2n_cipher_suite *original_suite = conn->secure->cipher_suite;
+    S2N_CIPHER_SUITE_WITH_LIMIT(key_limit_suite, original_suite, S2N_TEST_ENCRYPTION_LIMIT);
+
+    conn->secure->cipher_suite = &key_limit_suite;
+
+    RESULT_GUARD(s2n_send_and_recv_random_data(conn));
+    RESULT_GUARD(s2n_sanity_check_key_updates_sent(conn));
+
+    conn->secure->cipher_suite = original_suite;
+    return S2N_RESULT_OK;
+}
+
+static S2N_RESULT s2n_test_peer_requests(struct s2n_connection *conn)
+{
+    RESULT_GUARD_POSIX(s2n_example_negotiate(conn));
+
+    struct s2n_cipher_suite *original_suite = conn->secure->cipher_suite;
+    S2N_CIPHER_SUITE_WITH_LIMIT(key_limit_suite, original_suite, S2N_TEST_ENCRYPTION_LIMIT);
+
+    conn->secure->cipher_suite = &key_limit_suite;
+    if (conn->mode == S2N_CLIENT) {
+        RESULT_GUARD(s2n_set_key_update_request_for_testing(S2N_KEY_UPDATE_REQUESTED));
+    }
+
+    RESULT_GUARD(s2n_send_and_recv_random_data(conn));
+    RESULT_GUARD(s2n_sanity_check_key_updates_sent(conn));
+
+    conn->secure->cipher_suite = original_suite;
+    return S2N_RESULT_OK;
+}
+
+typedef S2N_RESULT (*s2n_test_scenario)(struct s2n_connection *conn);
+static S2N_RESULT s2n_run_self_talk_test(s2n_test_scenario scenario_fn)
+{
+    DEFER_CLEANUP(struct s2n_cert_chain_and_key *chain_and_key = NULL,
+            s2n_cert_chain_and_key_ptr_free);
+    RESULT_GUARD_POSIX(s2n_test_cert_chain_and_key_new(&chain_and_key,
+            S2N_DEFAULT_TEST_CERT_CHAIN, S2N_DEFAULT_TEST_PRIVATE_KEY));
+
+    DEFER_CLEANUP(struct s2n_config *config = s2n_config_new(),
+            s2n_config_ptr_free);
+    RESULT_GUARD_POSIX(s2n_config_set_unsafe_for_testing(config));
+    RESULT_GUARD_POSIX(s2n_config_set_cipher_preferences(config, "default_tls13"));
+    RESULT_GUARD_POSIX(s2n_config_add_cert_chain_and_key_to_store(config, chain_and_key));
+
+    DEFER_CLEANUP(struct s2n_test_io_pair io_pair = { 0 }, s2n_io_pair_close);
+    RESULT_GUARD_POSIX(s2n_io_pair_init_non_blocking(&io_pair));
+
+    pid_t client_pid = fork();
+    if (client_pid == 0) {
+        /* Suppress stdout.
+         * This only affects the new client process.
+         */
+        fclose(stdout);
+
+        struct s2n_connection *client = s2n_connection_new(S2N_CLIENT);
+        EXPECT_NOT_NULL(client);
+        EXPECT_SUCCESS(s2n_connection_set_config(client, config));
+
+        EXPECT_SUCCESS(s2n_io_pair_close_one_end(&io_pair, S2N_SERVER));
+        EXPECT_SUCCESS(s2n_connection_set_io_pair(client, &io_pair));
+
+        EXPECT_OK(scenario_fn(client));
+
+        EXPECT_SUCCESS(s2n_connection_free(client));
```

Comment:
I was initially using DEFER_CLEANUP on the connection, but that was causing one of our valgrind tests to somehow miss a free that happens in s2n_connection_free and fail. I wonder if it was because the connection went out of scope with exit instead of return? But DEFER_CLEANUP still worked for s2n_examples_test.c.

But anyway, valgrind passes without DEFER_CLEANUP.